src/diff_generate.c

```diff
@@ -273,7 +273,8 @@ static git_diff_delta *diff_delta__last_for_item(
 		break;
 	case GIT_DELTA_MODIFIED:
 		if (git_oid__cmp(&delta->old_file.id, &item->id) == 0 ||
-			git_oid__cmp(&delta->new_file.id, &item->id) == 0)
+		    (delta->new_file.mode == item->mode &&
```

Comment:
Do modes have to be completely the same? I get that it doesn't make sense in case one is a directory and the other is not. But what about when e.g. 644 changes to 755?

Reply:
I have to admit that I don't really understand this code.  It seems completely weird to me that something gets added to the diff list and then later removed.  But gdb shows the "mode" for a directory as 16384 -- that is, these are git modes, not filesystem modes.  And looking at all three of the callers, they only happen in the case where one side of the comparison is a directory. So I don't think mode changes are relevant here.  What do you think?

Reply:
Goodness, I don't know what this code is doing, either.

Reply:
If the question is "should X get removed from the list of diffs because it's actually identical to this other thing", I think a mode change should give the answer "no", because there is still a diff. 

(But, again, there's no case where this is called on two files; one arg is always a directory)